doc/doxygen/man/man3/Hmaps.3
.TH "Hmaps" 3 "Wed Oct 6 2021" "MAGRATHEA/PATHFINDER" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Hmaps \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <hmaps\&.h>\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "template<class Parameters , class Map > static void \fBReadParamFile\fP (Parameters &\fBparameters\fP, Map &parameter)"
.br
.RI "\fIRead parameter file\&. \fP"
.ti -1c
.RI "template<class Parameter , typename Integer , class Cones > static void \fBgetPixels_per_cone\fP (const Parameter &\fBparameters\fP, const Integer npix, std::vector< Integer > &pixel, std::vector< unsigned int > &ntrajectories, const unsigned int iconerank, const Cones &cones)"
.br
.RI "\fIGet targets\&. \fP"
.ti -1c
.RI "template<class Parameter , typename Integer , class Cones > static void \fBgetPixels_per_cone2\fP (const Parameter &\fBparameters\fP, const Integer npix, std::vector< Integer > &pixel, std::vector< unsigned int > &ntrajectories, const unsigned int iconerank, const Cones &cones)"
.br
.RI "\fIGet targets\&. \fP"
.ti -1c
.RI "template<class Parameter , class Octree , class Map , class Pixel , class Cosmology , class Point , typename Integer , typename Real > static void \fBFillMapPropagate\fP (const Parameter &\fBparameters\fP, const Integer ntrajectory, const Integer firsttrajectory, const Octree &octree, const Point &vobs, Map &map, const Integer nmaps, const Pixel &pixel, const Cosmology &cosmology, const \fBpoint\fP &observer, const Real length, const std::vector< Real > &z_stop_vec)"
.br
.RI "\fIProduce Healpix maps using bundles\&. \fP"
.ti -1c
.RI "template<class Parameter , class Octree , class Map , class Pixel , class Cosmology , class Point , typename Integer , typename Real > static void \fBFillMap\fP (const Parameter &\fBparameters\fP, const std::vector< std::string > &map_components, const std::vector< unsigned int > &index_components, const Integer ntrajectory, const Integer firsttrajectory, const Octree &octree, const Point &vobs, Map &map, const unsigned int nmaps, const Pixel &pixel, const Cosmology &cosmology, const \fBpoint\fP &observer, const Real length, const std::vector< Real > &interpRefvec, const std::vector< Real > &rhomo, const std::vector< Real > &thomo, const std::vector< Real > &lambdahomo, const std::vector< Real > &redshifthomo, const std::vector< Real > &ahomo)"
.br
.RI "\fIProduce Healpix maps\&. \fP"
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "template<class Parameter , class Octree , class Map , class Pixel , class Cosmology , class Point , typename Integer , typename Real > void Hmaps::FillMap (const Parameter &parameters, const std::vector< std::string > &map_components, const std::vector< unsigned int > &index_components, const Integerntrajectory, const Integerfirsttrajectory, const Octree &octree, const Point &vobs, Map &map, const unsigned intnmaps, const Pixel &pixel, const Cosmology &cosmology, const \fBpoint\fP &observer, const Reallength, const std::vector< Real > &interpRefvec, const std::vector< Real > &rhomo, const std::vector< Real > &thomo, const std::vector< Real > &lambdahomo, const std::vector< Real > &redshifthomo, const std::vector< Real > &ahomo)\fC [static]\fP"

.PP
Produce Healpix maps\&. Produce Healpix maps 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameter\fP Parameter type 
.br
\fIOctree\fP octree type 
.br
\fIMap\fP map type 
.br
\fIPixel\fP pixel type 
.br
\fICosmology\fP cosmology type 
.br
\fIPoint\fP Point type 
.br
\fIInteger\fP integer type 
.br
\fIReal\fP real/float type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIparameters\fP Parameter structure\&. 
.br
\fImap_components\fP Vector of strings containing the map types\&. 
.br
\fIindex_components\fP Vector containing the index relative to the map types\&. 
.br
\fIntrajectory\fP Number of pixels to be filled 
.br
\fIfirsttrajectory\fP first index for pixel for a given cone 
.br
\fIoctree\fP Octree 
.br
\fIvobs\fP Observer velocity 
.br
\fImap\fP Map 
.br
\fInmaps\fP Number of Healpix maps 
.br
\fIpixel\fP Pixel vector 
.br
\fIcosmology\fP Cosmological table 
.br
\fIobserver\fP Observer position 
.br
\fIlength\fP R\&.U to S\&.I units for length 
.br
\fIinterpRefvec\fP vector of interpolations at a given quantity for a given surface 
.br
\fIrhomo\fP vector of FLRW distance at the redshifts of interest 
.br
\fIthomo\fP vector of FLRW time at the redshifts of interest 
.br
\fIlambdahomo\fP vector of FLRW affine parameter at the redshifts of interest 
.br
\fIredshifthomo\fP vector of FLRW redshift at the redshifts of interest 
.br
\fIahomo\fP vector of FLRW scale factor at the redshifts of interest 
.RE
.PP

.SS "template<class Parameter , class Octree , class Map , class Pixel , class Cosmology , class Point , typename Integer , typename Real > void Hmaps::FillMapPropagate (const Parameter &parameters, const Integerntrajectory, const Integerfirsttrajectory, const Octree &octree, const Point &vobs, Map &map, const Integernmaps, const Pixel &pixel, const Cosmology &cosmology, const \fBpoint\fP &observer, const Reallength, const std::vector< Real > &z_stop_vec)\fC [static]\fP"

.PP
Produce Healpix maps using bundles\&. Produce Healpix maps using bundles at constant FLRW redshift 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameter\fP Parameter type 
.br
\fIOctree\fP octree type 
.br
\fIMap\fP map type 
.br
\fIPixel\fP pixel type 
.br
\fICosmology\fP cosmology type 
.br
\fIInteger\fP integer type 
.br
\fIReal\fP real/float type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIparameters\fP Parameter structure 
.br
\fIntrajectory\fP Number of pixels to be filled 
.br
\fIfirsttrajectory\fP first index for pixel for a given cone 
.br
\fIoctree\fP Octree 
.br
\fIvobs\fP Observer velocity 
.br
\fImap\fP Map 
.br
\fInmaps\fP Number of Healpix maps 
.br
\fIpixel\fP Pixel vector 
.br
\fIcosmology\fP Cosmological table 
.br
\fIobserver\fP Observer position 
.br
\fIlength\fP R\&.U to S\&.I units for length 
.br
\fIz_stop_vec\fP Vector containing redshifts at which we compute scalar quantities 
.RE
.PP

.SS "template<class Parameter , typename Integer , class Cones > void Hmaps::getPixels_per_cone (const Parameter &parameters, const Integernpix, std::vector< Integer > &pixel, std::vector< unsigned int > &ntrajectories, const unsigned inticonerank, const Cones &cones)\fC [static]\fP"

.PP
Get targets\&. Get targets inside cone\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameter\fP Parameter type 
.br
\fIInteger\fP Integer type 
.br
\fICones\fP Cones type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIparameters\fP Parameter structure\&. 
.br
\fInpix\fP Number of pixels\&. 
.br
\fIpixel\fP Vector of pixels inside cones that are treated by the same MPI task\&. 
.br
\fIntrajectories\fP Number of pixels per cone\&. 
.br
\fIiconerank\fP Number of the cone of interest\&. 
.br
\fICones\fP Geometry of all the cones\&. return Filled vector of targets inside cones for a given MPI task 
.RE
.PP

.SS "template<class Parameter , typename Integer , class Cones > void Hmaps::getPixels_per_cone2 (const Parameter &parameters, const Integernpix, std::vector< Integer > &pixel, std::vector< unsigned int > &ntrajectories, const unsigned inticonerank, const Cones &cones)\fC [static]\fP"

.PP
Get targets\&. Get targets inside cone\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameter\fP Parameter type 
.br
\fIInteger\fP Integer type 
.br
\fICones\fP Cones type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIparameters\fP Parameter structure\&. 
.br

src/hmaps.cpp
/* ******************************* HMAPS ******************************** */
/*////////////////////////////////////////////////////////////////////////////*/
// PROJECT :        PATHFINDER
// TITLE :          Hmaps
// DESCRIPTION :    Main function of the raytracer to create Healpix maps
// AUTHOR(S) :      Vincent Reverdy (vince.rev@gmail.com), Michel-Andrès Breton (michel-andres.breton@obspm.fr)
// CONTRIBUTIONS :  [Vincent Reverdy (2012-2013), Michel-Andrès Breton (2015-2021)]
// LICENSE :        CECILL-B License
/*////////////////////////////////////////////////////////////////////////////*/
/// \file           hmaps.cpp
/// \brief          Main function of the raytracer to create Healpix maps
/// \author         Vincent Reverdy (vince.rev@gmail.com), Michel-Andrès Breton (michel-andres.breton@obspm.fr)
/// \date           2012-2013, 2015-2021
/// \copyright      CECILL-B License
/*////////////////////////////////////////////////////////////////////////////*/



// ------------------------------ PREPROCESSOR ------------------------------ //
#include <ctime>
// Include C++
#include <iostream>
#include <iomanip>
#include <fstream>
#include <algorithm>
#include <utility>
#include <random>
#include <string>
#include <vector>
#include <deque>
#include <array>
#include <tuple>
#include <atomic>
#include <mutex>
// Include libs
#include <mpi.h>
// Include project
#include "magrathea/simplehyperoctree.h"
#include "magrathea/simplehyperoctreeindex.h"
#include "magrathea/hypersphere.h"
#include "magrathea/hypercube.h"
#include "magrathea/constants.h"
#include "magrathea/evolution.h"
#include "magrathea/timer.h"
#include "cone.h"
#include "utility.h"
#include "input.h"
#include "output.h"
#include "integrator.h"
#include "miscellaneous.h"
#include "hmaps.h"
// Octree
#ifdef VELOCITYFIELD
#include "gravity2.h" // with velocity slots
#else
#include "gravity.h"
#endif
// Include C++ HDF5 project
#include "TReadHDF5.h"

// Include HEALPIX
#include <chealpix.h>
#include <pointing.h>
#include <healpix_map.h>

// In miscellaneous.h we use std::sample. For versions of gcc lower than 7.x, please use std::experimental::sample. Check the Makefiles

// Misc
using namespace magrathea;
// -------------------------------------------------------------------------- //



// ---------------------------------- MAIN ---------------------------------- //
// Main function
/// \brief          Main function.
/// \details        Main function of the program. It takes a namelist as a 
///                 parameter and run the raytracing using MPI parallelization.
///                 To compile it, use make
///                 To run it, use:
///                 mpirun -np NTASKS ./raytracer raytracer.txt
/// \param[in]      argc Number of arguments.
/// \param[in]      argv List of arguments.
/// \return         Zero on success, error code otherwise.
int main(int argc, char* argv[])
{
    // Constants

    using integer = int;
    using uint = unsigned int;
    using real = double;
    using floating = float;
    using point = std::array<real, 3>;
    using position = std::ratio<0>;
    using extent = std::ratio<EXTENT>;
    using indexing = __uint128_t; 
#ifdef VELOCITYFIELD
    using element = std::pair<SimpleHyperOctreeIndex<indexing, 3>, Gravity<floating, 3> >;
#endif
    static constexpr uint zero = 0;
    static constexpr uint one = 1;
    static constexpr uint two = 2;
    static constexpr uint dimension = 3;
    static constexpr uint nreference = 5; // Used to set homogeneous octree
    static constexpr real rposition = static_cast<real>(position::num)/static_cast<real>(position::den);
    static constexpr point center({{rposition, rposition, rposition}});
    static constexpr real diameter = static_cast<real>(extent::num)/static_cast<real>(extent::den);
    static const std::string all = "all";
    static const std::string namelist = argc > 1 ? std::string(argv[1]) : std::string("raytracer.txt"); 

    // Parameters
    std::map<std::string, std::string> parameter;
    integer nthreads = std::thread::hardware_concurrency();
    integer ntasks = nthreads*zero;
    integer rank = zero;

    // Message passing interface
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &ntasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    // Read parameter file
    Miscellaneous::TicketizeFunction(rank, ntasks, [=, &parameter]{parameter = Input::parse(namelist);});
    // Convert strings and put it in struct
    Hmaps::ReadParamFile(parameters, parameter);
    // Initialization
    FileList conefile(parameters.conefmt, zero, parameters.ncones, zero, parameters.conedir);
    SimpleHyperOctree<real, SimpleHyperOctreeIndex<indexing, dimension>, Gravity<floating, dimension>, dimension, position, extent> octree;
    HyperSphere<dimension, point> sphere(center, diameter/two);
    std::vector<Cone<point> > cone(parameters.ncones);
    std::vector<Cone<point> > coneIfRot(parameters.ncones);
    std::array<std::vector<real>, two+two> cosmology;
    Photon<real, dimension> photon;
    Evolution<Photon<real, dimension> > reference;
    real h = zero;
    real omegam = zero;
    real lboxmpch = zero;
    real thetay(0), thetaz(0);
    std::array< std::array< double, 3 >, 3 > rotm1 = {{zero}};
    const point vobs0 = {0,0,0}; // No peculiar velocity for homogeneous quantities
    std::mt19937 engine1(parameters.seed > zero ? parameters.seed+rank : std::random_device()());

    if(rank == 0) 
	std::cout<<"#### MAGRATHEA_PATHFINDER "<<std::endl; 
    // Generate cones
    if(parameters.isfullsky){
	if(rank == 0){
	    std::cout<<"## Fullsky cone"<<std::endl;
	}
        Miscellaneous::GenerateFullskyCones(parameters.ncones, cone, coneIfRot, sphere);
    } else{
	if(rank == 0){
	    std::cout<<"## Narrow cone"<<std::endl;
	}
        Miscellaneous::GenerateNarrowCones(parameters, cone, coneIfRot, sphere, rotm1, thetay, thetaz);
    }
    // Read cosmology
    cosmology = Input::acquire(parameters, h, omegam, lboxmpch);
    if(rank == 0){
        std::cout<<"## Parameter file "<<parameters.paramfile<<std::endl;
        std::cout<<"## Evolution file "<<parameters.evolfile<<std::endl;
        std::cout<<"## Cosmology : h = "<<h<<", omega_m = "<<omegam<<", rhoch2 = "<<parameters.rhoch2<<std::endl;
        std::cout<<"## Box length : "<<lboxmpch<<" Mpc/h, 1 Mpc = "<<parameters.mpc<<" meters "<<std::endl;
        std::cout<<"## Coarse level : "<<parameters.ncoarse<<", number of cones = "<<parameters.ncones<<std::endl;
    }
    const double length = lboxmpch*parameters.mpc/h;

    // Construct homogeneous tree 
    Input::homogenize(octree.assign(nreference, zero));
    reference.append(Integrator::launch(center[zero], center[one], center[two], center[zero]+diameter/two, center[one], center[two]));
    // Propagate a photon in a very refined homogeneous FLRW metric
    Integrator::integrate<-1>(reference, cosmology, octree, vobs0, length, EXTENT*std::pow(two, static_cast<uint>(std::log2(std::get<0>(cosmology).size()/std::pow(two, nreference)+one)+one)+one));
    // Correct cosmology with photon
    cosmology = Input::correct(cosmology, reference);
    octree.fullclear();

    // Execution
    if(rank == 0)
        std::cout<<"# Compute Healpix maps, nside = "<<parameters.nside<<std::endl;
    std::vector<real> z_stop_vec(parameters.nb_z_maps);
    // Get redshift values from redshift range and number of redhsifts
    for(uint i = 0; i < parameters.nb_z_maps; i++){
	z_stop_vec[i] = parameters.z_stop_min + i*(parameters.z_stop_max - parameters.z_stop_min)/(parameters.nb_z_maps - (parameters.nb_z_maps > 1));
    }
    // Get number of pixels in a fullsky map from with nside
    long npix = nside2npix(parameters.nside);

    std::vector<std::string> map_components;
    // Tokenize map types to put in vector
    Miscellaneous::Tokenize(parameters.map_components, map_components, ", ");
    std::vector<uint> index_components(map_components.size());
    uint nmaps(0);

    if(rank == 0)
        std::cout<<"Components of Healpix maps:"<<std::endl;
    // For each map type, assign a unique index. 'Lensing' types produce multiple maps
    for(uint i = 0; i < map_components.size(); i++){
	if(map_components[i] == "lensing"){
	    index_components[i] = 0;
	    if(parameters.jacobiantype=="bundle"){
	        if(rank == 0){
	            std::cout<<"Index: "<<nmaps<<", component: kappa ('lensing')"<<std::endl;
	            std::cout<<"Index: "<<nmaps+1<<", component: gamma1 ('lensing')"<<std::endl;
	            std::cout<<"Index: "<<nmaps+2<<", component: gamma2 ('lensing')"<<std::endl;
	            std::cout<<"Index: "<<nmaps+3<<", component: w ('lensing')"<<std::endl;
	            std::cout<<"Index: "<<nmaps+4<<", component: inverse magnification ('lensing')"<<std::endl;	
	        }
	        nmaps += 4;
	    } else if(parameters.jacobiantype=="infinitesimal"){
	        if(rank == 0){
	            std::cout<<"Index: "<<nmaps<<", component: kappa ('lensing')"<<std::endl;
	            std::cout<<"Index: "<<nmaps+1<<", component: gamma1 ('lensing')"<<std::endl;
	            std::cout<<"Index: "<<nmaps+2<<", component: gamma2 ('lensing')"<<std::endl;
	            std::cout<<"Index: "<<nmaps+3<<", component: inverse magnification ('lensing')"<<std::endl;	
	        }
	        nmaps += 3;
	    } else{
		std::cout<<"# WARNING: With 'lensing', please choose jacobiantype 'bundle' or 'infinitesimal'"<<std::endl;
	        std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
	        std::terminate();
	    }   
	} else if(map_components[i] == "lensing_born"){
	    index_components[i] = 1;
	    if(rank == 0){
	        std::cout<<"Index: "<<nmaps<<", component: kappa ('lensing_born') WARNING: with lensing_born we compute the jacobian matrix with an infinitesimal beam (no bundle method available)"<<std::endl;
	        std::cout<<"Index: "<<nmaps+1<<", component: gamma1 ('lensing_born')"<<std::endl;
	        std::cout<<"Index: "<<nmaps+2<<", component: gamma2 ('lensing_born')"<<std::endl;
	        std::cout<<"Index: "<<nmaps+3<<", component: inverse magnification ('lensing_born')"<<std::endl;	
	    }
	    nmaps += 3;    
	} else if(map_components[i] == "deflection"){
	    index_components[i] = 13;
	    if(rank == 0){
	        std::cout<<"Index: "<<nmaps<<", component: alpha1 ('deflection') [dphi in spherical coordinates]"<<std::endl;
	        std::cout<<"Index: "<<nmaps+1<<", component: alpha2 ('deflection') [dtheta in spherical coordinates]"<<std::endl;
	    }
	    nmaps++;
	} else{
	    if(rank == 0)
	        std::cout<<"Index: "<<nmaps<<", component: '"<<map_components[i]<<"'"<<std::endl;
	    if(map_components[i] == "dr")
	  	index_components[i] = 2;
	    else if(map_components[i] == "dl")
	  	index_components[i] = 3;
	    else if(map_components[i] == "dt")
	  	index_components[i] = 4;
	    else if(map_components[i] == "da")
	  	index_components[i] = 5;
	    else if(map_components[i] == "dz")
	  	index_components[i] = 6;
	    else if(map_components[i] == "ds")
	  	index_components[i] = 7;
	    else if(map_components[i] == "isw")
	  	index_components[i] = 8;
	    else if(map_components[i] == "dens")
	  	index_components[i] = 9;
	    else if(map_components[i] == "dens_max")
	  	index_components[i] = 10;
	    else if(map_components[i] == "steps")
	  	index_components[i] = 11;
	    else if(map_components[i] == "phi")
	  	index_components[i] = 12;
	    else{
		std::cout<<"# WARNING: Need good name for map_components"<<std::endl;
	        std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
	        std::terminate();
	    }
	} 
	nmaps++;
    }
    // Create vector of Healpix maps
    std::vector <Healpix_Map<float> > map(parameters.nb_z_maps*nmaps);
    // Initialise maps
    for( uint i = 0; i < parameters.nb_z_maps*nmaps; i++){
        map[i].SetNside((int) parameters.nside, (Healpix_Ordering_Scheme) 0);
        map[i].fill(0.);
    }

#ifdef VERBOSE
    if(rank == 0)
        std::cout<<"# Pixels attribution"<<std::endl;
#endif
    std::vector<long> pixel;
    std::vector<uint> ntrajectories;
    // Assign pixels to cones
    for (uint iconerank = zero; iconerank < parameters.ncones; ++iconerank) {
	if (iconerank%static_cast<uint>(ntasks) == static_cast<uint>(rank)) {
	    Hmaps::getPixels_per_cone2(parameters, npix, pixel, ntrajectories, iconerank, coneIfRot);
	} 
    } 
    pixel.shrink_to_fit();
#ifdef VERBOSE
    std::cout<<"# Pixels distributed : rank "<<rank<<" takes care of "<<pixel.size()<<" pixels"<<std::endl;
#endif
    point observer; 
    observer[0] = 0;
    observer[1] = 0;
    observer[2] = 0;
    std::vector<real> interpRefvec(parameters.nb_z_maps);
    std::vector<double> thomo(parameters.nb_z_maps), rhomo(parameters.nb_z_maps), lambdahomo(parameters.nb_z_maps), redshifthomo(parameters.nb_z_maps), ahomo(parameters.nb_z_maps);

    // For each redshift, estimate the homogeneous quantities [scale factor, redshift, comoving distance, conformal time, affine parameter]
    for(unsigned int i = 0; i < parameters.nb_z_maps; i++){
	photon.redshift() = z_stop_vec[i];
	const unsigned long int marked = std::distance(std::begin(reference), std::upper_bound(std::begin(reference), std::end(reference), photon, [](const Photon<double, 3>& first, const Photon<double, 3>& second){return first.redshift() < second.redshift();}));
	const unsigned long int firstid = marked - (marked > 0);
	const double f = (reference[firstid+1].redshift() - photon.redshift())/(reference[firstid+1].redshift() - reference[firstid].redshift());
	thomo[i] = reference[firstid].t()*f + reference[firstid+1].t()*(1-f);
	rhomo[i] = reference[firstid].chi()*f + reference[firstid+1].chi()*(1-f);
	lambdahomo[i] = reference[firstid].lambda()*f + reference[firstid+1].lambda()*(1-f);
	ahomo[i] = 1./(1.+z_stop_vec[i]);
	redshifthomo[i] = z_stop_vec[i];
	if(parameters.stop_ray == "t"){
	    interpRefvec = thomo;
	} else if (parameters.stop_ray == "r"){
	    interpRefvec = rhomo;
	} else if (parameters.stop_ray == "lambda"){
	    interpRefvec = lambdahomo;
	} else if (parameters.stop_ray == "a"){
	    interpRefvec = ahomo;
	} else if (parameters.stop_ray == "redshift"){
	    interpRefvec = z_stop_vec;
	} else if (parameters.stop_ray == "s"){
	    interpRefvec = rhomo;
	} else{
	    std::cout<<"Need good interpolation 't, r, lambda, a, redshift or ell'"<<std::endl;
	    std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
	    std::terminate();
	}

	if(rank == 0){
	    std::cout<<std::setprecision(17)<<"# Homogeneous quantities : z = "<<redshifthomo[i]<<" a = "<<ahomo[i]<<" t = "<<thomo[i]<<" chi = "<<rhomo[i]<<" lambda = "<<lambdahomo[i]<<std::endl;
	}
    }

    reference.fullclear();

    // Fill map
#ifndef VELOCITYFIELD
    // No velocity field calculation
    uint firsttrajectory(0);
    long itraj = 0;
    for (uint icone = zero; icone < parameters.ncones; ++icone) {
        if (icone%static_cast<uint>(ntasks) == static_cast<uint>(rank)) {
#ifdef VERBOSE
            std::cout<<"# Rank "<<rank<<" using cone "<<icone<<std::endl;
#endif
	    // Load octree from binary file
	    Miscellaneous::loadOctree(icone, octree, conefile);
#ifdef VERBOSE
   	    std::cout<<"# Rank "<<rank<<" using cone "<<icone<<", now computing Healpix maps"<<std::endl;
#endif                  
	    // Fill Healpix maps with ray-tracing routines    
	    Hmaps::FillMap(parameters, map_components, index_components, ntrajectories[itraj], firsttrajectory, octree, vobs0, map, nmaps, pixel, cosmology, observer, length, interpRefvec, rhomo, thomo, lambdahomo, redshifthomo, ahomo);
	    firsttrajectory += ntrajectories[itraj];
	    itraj++;
	}   
    } 
#else
     // Account for velocity field
#ifdef VERBOSE
    if(rank == 0)
	std::cout<<"# Maps at inhomogeneous redshift(s). Velocity field computed with "<<parameters.velocity_field<<std::endl;
#endif
    const point vobs = {parameters.v0x, parameters.v0y, parameters.v0z};
    std::size_t found;
    std::vector<std::string> filelistprior, shellList;

    // Get all filenames in particle directory
    Miscellaneous::getFilesinDir(parameters.partdir, filelistprior);
    for(uint ifiling = 0; ifiling < filelistprior.size(); ++ifiling){
	// Only keep files with 'shell' in their name
        found = filelistprior[ifiling].find("shell");
	if (found!=std::string::npos){
	    shellList.push_back(parameters.partdir+filelistprior[ifiling]);
	}
    }
    Miscellaneous::fullclear_vector(filelistprior);
    // Shuffle filenames so that different MPI tasks read different files
    std::shuffle(std::begin(shellList), std::end(shellList), engine1);

    uint firsttrajectory(0);
    long itraj = 0;
    for (uint icone = zero; icone < parameters.ncones; ++icone) {
	if (icone%static_cast<uint>(ntasks) == static_cast<uint>(rank)) {
#ifdef VERBOSE
	    std::cout<<"# Rank "<<rank<<" using cone "<<icone<<", getting particles..."<<std::endl;
#endif
	    // Load octree from binary file
	    Miscellaneous::loadOctree(icone, octree, conefile);
	    // WARNING ! Set rho (density) to zero in each cell of the octree; 
	    Utility::parallelize(octree.size(), [=, &octree](const uint i){std::get<1>(octree[i]).rho() = 0;});
	    // Read Particle files to compute the velocity field in each AMR cell
	    for(uint ifile = 0; ifile < shellList.size(); ifile++){
		std::vector<float> pos_part, vel_part; 
		//Miscellaneous::TicketizeFunction(rank, ntasks, [=, &shellList, &pos_part, &vel_part]{ // Uncomment in case of IO problems
		// Get particles from files
	    	Hmaps::fill_particles_vectors(parameters, rotm1, cone[icone], shellList[ifile], pos_part, vel_part, z_stop_vec, thetay, thetaz); 
		//});   
		if(pos_part.size() > 0){
#ifdef VERBOSE
       		    std::cout<<"# Rank "<<rank<<" using cone "<<icone<<", file : "<<ifile<<"/"<<shellList.size()<<" npart "<<pos_part.size()/3<<" : "<<shellList[ifile]<<std::endl;
#endif
		    // From particles, compute the velocity field with CIC or TSC scheme
		    if(parameters.velocity_field == "cic"){
	    	        Hmaps::CreateOctreeVelocityWithCIC(octree, pos_part, vel_part); // octree with velocity field using CIC interpolation 
#ifdef VERBOSE
		        std::cout<<"# Rank "<<rank<<" using cone "<<icone<<", Octree created with CIC"<<std::endl;
#endif
		    } else if(parameters.velocity_field == "tsc"){
	    	        Hmaps::CreateOctreeVelocityWithTSC(octree, pos_part, vel_part); // octree with velocity field using TSC interpolation 
#ifdef VERBOSE
		        std::cout<<"# Rank "<<rank<<" using cone "<<icone<<", Octree created with TSC"<<std::endl;
#endif
		    } else{
    		        std::cout<<"No velocity field computed, please type cic or tsc"<<std::endl;
			std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
			std::terminate();
		    }
		}
	    }		       
	    // Finalize
	    const unsigned int lvlmax = (std::get<0>(*std::max_element(std::begin(octree), std::end(octree), [](const element& x, const element& y){return std::get<0>(x).level() < std::get<0>(y).level();})).level()); 
	    const unsigned int lvlmin = (std::get<0>(*std::min_element(std::begin(octree), std::end(octree), [](const element& x, const element& y){return std::get<0>(x).level() < std::get<0>(y).level();})).level());
	    // Normalise velocity field with mass
	    Utility::parallelize(octree.size(), [=, &octree](const uint i){
		double mass = std::get<1>(octree[i]).rho();   
		mass = mass + (mass == 0);
		std::get<1>(octree[i]).vx() /=  mass;   
		std::get<1>(octree[i]).vy() /=  mass;   
		std::get<1>(octree[i]).vz() /=  mass; 
	    });
	    // If the density is zero in one cell, get value from parent
	    for (unsigned int ilvl = lvlmin+1; ilvl <= lvlmax; ilvl++) {
		Utility::parallelize(octree.size(), [=, &octree](const uint i){
		    if(std::get<0>(octree[i]).level() == ilvl){
			Gravity<floating, 3> data;
			if (!std::isnormal(std::get<1>(octree[i]).rho())) {
			    data = std::get<1>(*octree.find(std::get<0>(octree[i]).parent()));
		 	    std::get<1>(octree[i]).vxyz() = {data.vx(), data.vy(), data.vz()};
			}
		    }
		});
	    }   
			
	    // Fill map
#ifdef VERBOSE
	    std::cout<<"# Rank "<<rank<<" using cone "<<icone<<", now computing Healpix maps, ntrajectory = "<<ntrajectories[itraj]<<std::endl;
#endif
	    // Fill Healpix maps with ray-tracing routines    
	    Hmaps::FillMap(parameters, map_components, index_components, ntrajectories[itraj], firsttrajectory, octree, vobs, map, nmaps, pixel, cosmology, observer, length, interpRefvec, rhomo, thomo, lambdahomo, redshifthomo, ahomo);		
#ifdef VERBOSE
	    std::cout<<"# Rank "<<rank<<" using cone "<<icone<<", Healpix maps computed"<<std::endl;
#endif
	    firsttrajectory += ntrajectories[itraj];
	    itraj++;
	}
    }
#endif
	    
#ifdef VERBOSE
    std::cout<<"# Rank "<<rank<<" : Healpix maps filled, now waiting for other procs to finalize "<<std::endl;
#endif 
    Miscellaneous::fullclear_vector(pixel);
    octree.fullclear();
    // Wait for all tasks
    MPI_Barrier(MPI_COMM_WORLD);

    // Finalisation
    float * mapdata1;
    float * mapdata2;
    mapdata1 = (float *) malloc(npix * sizeof(float));
    mapdata2 = (float *) malloc(npix * sizeof(float));
    const char coordsys = 'G';
    // Create vector of Healpix maps
    Healpix_Map<float>  map_tmp;
    map_tmp.SetNside((int) parameters.nside, (Healpix_Ordering_Scheme) 0);
    int my_iz(-1), my_i(-1);

    // Loop over all the maps
    for(uint iz = 0; iz < parameters.nb_z_maps; iz++){
        for(uint i = 0; i < nmaps; i++){
	    // Copy map in temporary array
            map[nmaps*iz + i].Map().copyToPtr<float>(mapdata1);
            MPI_Barrier(MPI_COMM_WORLD);
	    // Add contribution from every MPI task mapdata1 into the array mapdata2
	    const int icone = (nmaps*iz + i) % static_cast<uint>(ntasks);
	    MPI_Reduce((void*) mapdata1, (void*) mapdata2, (int) npix, MPI_FLOAT, MPI_SUM, icone, MPI_COMM_WORLD);
	    if(rank == icone){
		my_iz = iz;
		my_i = i;
	    }
	    // Write the outputs if all the tasks contain a full map, or if the total number of maps are contained
	    //std::cout<<"rank = "<<rank<<" map "<<nmaps*iz + i<<" ntasks "<<ntasks<<" ratio "<<(nmaps*iz + i) % static_cast<uint>(ntasks)<<" max "<<static_cast<uint>(nmaps*parameters.nb_z_maps - 1)<<" my "<<my_iz<<" "<<my_i<<std::endl;
	    if (((nmaps*iz + i) % static_cast<uint>(ntasks) == static_cast<uint>(ntasks - 1)) | ((nmaps*iz + i) == static_cast<uint>(nmaps*parameters.nb_z_maps - 1)) ) {
		// Select tasks
		if(my_iz != -1){
	            // Clean the map
 	            for (long pix = 0; pix < npix; pix++){
		        // If pixel is not normal (i.e. zero, infinity or NaN), then put undefined value 
	                if (!std::isnormal(mapdata2[pix])){
		            mapdata2[pix] = Healpix_undef; 
		        }
	            }	
#ifdef VERBOSE
	            std::cout << " #-- Writing output with rank = "<<rank<<" redshift = "<<z_stop_vec[my_iz]<<" id = "<<my_i<< std::endl;
#endif
	            std::string outputfilename; 
		    outputfilename = parameters.outputdir + parameters.outputprefix + "_nside" + std::to_string(parameters.nside) + "_" + std::to_string(my_i) +"_z_"+std::to_string(z_stop_vec[my_iz])+"_"+parameters.stop_ray+".fits";
		    // Write Healpix map
	            write_healpix_map(mapdata2, parameters.nside, outputfilename.data(), 0, &coordsys);
		    my_iz = -1;
		}
	    }
        }
    }

    free(mapdata1);
    free(mapdata2);
    
    // Finalization
    MPI_Barrier(MPI_COMM_WORLD);
    MPI_Finalize();
    if(rank == 0)
        std::cout<<"# Run completed !"<<std::endl;
    return 0;
}
// -------------------------------------------------------------------------- //



/*////////////////////////////////////////////////////////////////////////////*/
